File: medium/Two-Pointers/valid-palindrome/valid_palindrome.cpp
#include <string>
#include <cctype>
using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        /*
         * Check if the string is a valid palindrome.
         * Approach: Two Pointers
         * Time: O(n)
         * Space: O(1)
         */
        // Initialize left and right pointers
        int left = 0;
        int right = s.size() - 1;

        while (left < right) {
            // Skip non-alphanumeric characters on left
            while (left < right && !isalnum(s[left])) left++;
            // Skip non-alphanumeric characters on right
            while (left < right && !isalnum(s[right])) right--;

            // Compare characters in lowercase
            if (tolower(s[left]) != tolower(s[right])) {
                return false;
            }

            // Move pointers towards center
            left++;
            right--;
        }
        return true;
    }
};

// Test
#include <iostream>
int main() {
    Solution sol;
    string s = "A man, a plan, a canal: Panama";
    cout << sol.isPalindrome(s) << endl;
    return 0;
}

File: medium/Two-Pointers/valid-palindrome/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        """
        Check if the string is a valid palindrome.
        Approach: Two Pointers
        Time: O(n)
        Space: O(1)
        """
        # Initialize left and right pointers
        left = 0
        right = len(s) - 1

        while left < right:
            # Skip non-alphanumeric characters on left
            while left < right and not s[left].isalnum():
                left += 1
            # Skip non-alphanumeric characters on right
            while left < right and not s[right].isalnum():
                right -= 1
            
            # Compare characters in lowercase
            if s[left].lower() != s[right].lower():
                return False
            
            # Move pointers towards center
            left += 1
            right -= 1
        
        return True

if __name__ == "__main__":
    sol = Solution()
    print(sol.isPalindrome("A man, a plan, a canal: Panama"))  # True
